clients/client-url-connection/src/main/java/io/techery/janet/http/UrlConnectionClient.java

```diff
@@ -101,28 +100,34 @@ Response readResponse(HttpURLConnection connection) throws IOException {
                 headers.add(new Header(name, value));
             }
         }
-
-        String mimeType = connection.getContentType();
-        InputStream stream;
-        if (status >= 400) {
-            stream = connection.getErrorStream();
-        } else {
-            stream = connection.getInputStream();
-        }
-        ActionBody responseBody = new BytesArrayBody(mimeType, streamToBytes(stream));
+        ActionBody responseBody = new ResponseActionBody(connection);
         return new Response(connection.getURL().toString(), status, reason, headers, responseBody);
     }
 
-    public static byte[] streamToBytes(InputStream stream) throws IOException {
-        ByteArrayOutputStream baos = new ByteArrayOutputStream();
-        if (stream != null) {
-            byte[] buf = new byte[BUFFER_SIZE];
-            int r;
-            while ((r = stream.read(buf)) != -1) {
-                baos.write(buf, 0, r);
+    private static class ResponseActionBody extends ActionBody {
+
+        private HttpURLConnection body;
+
+        public ResponseActionBody(HttpURLConnection body) {
+            super(body.getContentType());
+            this.body = body;
+        }
+
+        @Override public long length() {
+            return body.getContentLengthLong();
+        }
+
+        @Override public InputStream getContent() throws IOException {
+            if (body.getResponseCode() < 400) {
```

Comment:
Change the condition of checking success response code to right way covering all possible HTTP codes.